consensus/hotstuff/timeoutcollector/aggregation.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		return a.TotalWeight(), fmt.Errorf("couldn't verify signature from %s: %w", signerID, err)
+	}
+	if !valid {
+		return a.TotalWeight(), fmt.Errorf("invalid signature from %s: %w", signerID, model.ErrInvalidSignature)
+	}
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if _, duplicate := a.idToSignature[signerID]; duplicate {
+		return a.totalWeight, model.NewDuplicatedSignerErrorf("signature from %v was already added", signerID)
+	}
+
+	a.idToSignature[signerID] = sigInfo{
+		sig:        sig,
+		highQCView: highestQC.View,
+	}
+	a.totalWeight += info.weight
+
+	if a.highestQC == nil || a.highestQC.View < highestQC.View {
+		a.highestQC = highestQC
+	}
+
+	return a.totalWeight, nil
+}
+
+func (a *TimeoutSignatureAggregator) hasSignature(singerID flow.Identifier) bool {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+	_, found := a.idToSignature[singerID]
+	return found
+}
+
+// TotalWeight returns the total weight presented by the collected signatures.
+// The function is thread-safe
+func (a *TimeoutSignatureAggregator) TotalWeight() uint64 {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+	return a.totalWeight
+}
+
+// Aggregate aggregates the signatures and returns the aggregated signature.
+// The function performs a final verification of aggregated
+// signature. Caller can be sure that resulting signature is valid.
+// Expected errors during normal operations:
+//  - model.InsufficientSignaturesError if no signatures have been added yet
+// This function is thread-safe
+//
+func (a *TimeoutSignatureAggregator) Aggregate() (*flow.TimeoutCertificate, error) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	sharesNum := len(a.idToSignature)
+	if sharesNum == 0 {
+		return nil, model.NewInsufficientSignaturesErrorf("cannot aggregate an empty list of signatures")
+	}
+	pks := make([]crypto.PublicKey, 0, sharesNum)
+	signatures := make([]crypto.Signature, 0, sharesNum)
+	hashers := make([]hash.Hasher, 0, sharesNum)
+	signers := make([]flow.Identifier, 0, sharesNum)
+	messages := make([][]byte, 0, sharesNum)
+	highQCViews := make([]uint64, 0, sharesNum)
+	for id, info := range a.idToSignature {
+		msg := verification.MakeTimeoutMessage(a.view, info.highQCView)
+		signatures = append(signatures, info.sig)
+		signers = append(signers, id)
+		messages = append(messages, msg)
+		pks = append(pks, a.idToInfo[id].pk)
+		hashers = append(hashers, a.hasher)
+		highQCViews = append(highQCViews, info.highQCView)
+	}
+
+	aggSignature, err := crypto.AggregateBLSSignatures(signatures)
+	if err != nil {
+		// invalidInputsError for:
+		//  * empty `signatures` slice, i.e. sharesNum == 0, which we exclude by earlier check
+		//  * if some signature(s), included via TrustedAdd, could not be decoded
+		if crypto.IsInvalidInputsError(err) {
+			return nil, model.NewInvalidSignatureIncludedErrorf("signatures with invalid structure were included via TrustedAdd: %w", err)
+		}
+		return nil, fmt.Errorf("BLS signature aggregation failed: %w", err)
```

Comment:
agree with [Tarak's comment](https://github.com/onflow/flow-go/pull/2243#discussion_r861228522): 
* As we don't have the `TrustedAdd` method, this is _not_ an expected failure path anymore, but a symptom of an internal bug. Hence, no need to provide a sentinel to the higher logic. 

```suggestion
		// invalidInputsError for:
		//  * empty `signatures` slice, i.e. sharesNum == 0, which we exclude by earlier check
		//  * if some signature(s) could not be decoded, which should be impossible since we check all signatures before adding them
		// Hence, any error here is a symptom of an internal bug
		return nil, fmt.Errorf("unexpected internal error during BLS signature aggregation: %w", err)
```